### DISCUSSION TITLE
How would Matrix support be with this?

### DISCUSSION THREAD

**Discussion by nyabinary** [mycrl/turn-rs#96] [General]
Does turn-rs support the necessary features for Matrix calls?

**Comment by mycrl** [comment]
> Does turn-rs support the necessary features for Matrix calls?

Not necessarily, I don't use matrix.

However, it seems that matrix is not currently supported because I looked at the documentation for matrix and it uses the turn rest api to interact with the turn servers. turn-rs does not currently follow the implementation specification for the turn rest api because that protocol only provides user authentication functionality.

The reason I didn't choose to implement that protocol in the first place is that it only exists as an authentication api, but turn-rs requires a lot of other functionality via the http api, so I implemented the external api from scratch as I wanted;

Don't be disappointed though, it's easy to support this rfc and you've reminded me that if other services want to interact with the turn servers they have no choice but to follow the rfc, so I'll be providing support for this rfc.

The update is expected to be released in the next few days, as the changes are actually quite minor.

RFC: https://datatracker.ietf.org/doc/html/draft-uberti-behave-turn-rest-00

**Reply by nyabinary** [reply]
Very nice to see support is coming up :)

**Reply by nyabinary** [reply]
Any ETA btw for the new release with this :3

**Reply by mycrl** [reply]
> Any ETA btw for the new release with this :3

I'm in the process of writing tests for checking that the newly added functionality works, and may need your help.
I would like to actually integrate with matrix to check if the functionality works if you can help me.

**Reply by mycrl** [reply]
@nyabinary 

I've released a test version, I'm not sure if this version works correctly because of the lack of integration with the actual service, it would be best if you could help, if there are no issues with the feedback from your end I'll release the official version.

https://github.com/mycrl/turn-rs/releases/tag/v3.1.0-beta

**Reply by nyabinary** [reply]
> @nyabinary
> 
> I've released a test version, I'm not sure if this version works correctly because of the lack of integration with the actual service, it would be best if you could help, if there are no issues with the feedback from your end I'll release the official version.
> 
> https://github.com/mycrl/turn-rs/releases/tag/v3.1.0-beta

Seems to be working for me, :P I don't know how to make a test exactly, though maybe it's better to ask the Matrix devs for that?
https://matrix.to/#/#matrix-dev:matrix.org

**Comment by mycrl** [comment]
> Does turn-rs support the necessary features for Matrix calls?

https://github.com/mycrl/turn-rs/releases/tag/v3.1.0

**Comment by mycrl** [comment]
And, regarding how to replace coturn, just provide the static key in the configuration of turn-rs, the configuration item `static_auth_secret` is the same name as coturn.

**Comment by rzerres** [comment]
This is great news. I did pack this crate for arch linux. Going to test it with my own homeserver.
I did had a look a couple of weeks ago. Looks so promising, but at that time, it was missing static.auth.secret.
No it should work and allow me to utilise the turn communication via `conduit`, a rust implemented matrix server.
Thank you Lazy Panda. Well done!

**Reply by rzerres** [reply]
Just a short feedback: It is working for me flawlessly!

At least the legacy my tests calls with `Element` are working.

 In a next step I have to implement the `Element Call` infrastructur components (Livekit, element-call, Webhooks). With the latest `Element X`  (as well as Rust-SDK based Clients, e.g. ShildiNext) apps it is possible to change the default connector (default: https://call.element.io). So it is possible to instantiate a `company local` SFU.

**Reply by mycrl** [reply]
I'm glad to help you, if there are other problems you can always feedback, happy using!

**Reply by nyabinary** [reply]
Hmm does turn-rs only support insecure turn?

**Reply by mycrl** [reply]
> Hmm does turn-rs only support insecure turn?

Yes, because turn-rs was originally created for a webrtc application, and webrtc itself has transport layer encryption, and within turn-rs, an unencrypted transport layer, encrypted data is transmitted, so turn-rs doesn't need to be encrypted anymore. But the turn server is a general purpose service, so turn doesn't limit the usage scenarios, just because I'm an audio/video rtc developer and the purpose of creating this project depends on my own preferences at the time.

**Reply by nyabinary** [reply]
Any plans on adding secure turn in the future?

**Reply by mycrl** [reply]
> Any plans on adding secure turn in the future?

The dtls implementation for encrypted udp is a bit more complex, and if it's just a matter of achieving the goal of being able to work, then some of the dtls libraries in crates.io can be used, but as far as I know there is no mature and well-performing implementation for crates.io. The most mature implementation is https://crates.io/crates/webrtc-dtls, but I've looked at the source code for this library before and the implementation is very bland (meaning there is a lot of room for optimization).
Therefore, if I want to implement encryption, I need to implement my own set of dtls logic, mainly optimized for turn-rs usage scenarios, which will require more work.
However, I will add support for transport layer encryption to the roadmap, but not too soon. Whether I fork webrtc-dtls or implement my own dtls, the amount of work won't be too small, because turn-rs is characterized by “great performance”, so I won't just find a third-party library to implement this feature.
Transport layer encryption should be added in the next big release, but it will take a bit longer.
